win32_deps_build.sh

```diff
@@ -0,0 +1,327 @@
+sudo python3 -m pip install cython
+
+cd $depsSrcDir
+if [[ ! -d $zlibDir ]]; then
+    git clone https://github.com/madler/zlib
+fi
+cd zlib
+# Apparently the configure script is broken...
+sed -e s/"PREFIX ="/"PREFIX = x86_64-w64-mingw32-"/ -i win32/Makefile.gcc
+_make -f win32/Makefile.gcc
+_make BINARY_PATH=$zlibDir \
+     INCLUDE_PATH=$zlibDir/include \
+     LIBRARY_PATH=$zlibDir/lib \
+     SHARED_MODE=1 \
+     -f win32/Makefile.gcc install
+
+cd $depsToolsetDir
+if [[ ! -d $lz4Dir ]]; then
+    git clone https://github.com/lz4/lz4
+fi
+cd lz4
+git checkout $lz4Tag
+_make BUILD_STATIC=no CC=x86_64-w64-mingw32-gcc \
+      DLLTOOL=x86_64-w64-mingw32-dlltool OS=Windows_NT
+
+cd $depsSrcDir
+if [[ ! -d $sslSrcDir ]]; then
+    curl "https://www.openssl.org/source/openssl-${sslVersion}.tar.gz" | tar xz
+fi
+cd $sslSrcDir
+mkdir -p $sslDir
+CROSS_COMPILE="x86_64-w64-mingw32-" ./Configure \
+    mingw64 shared --prefix=$sslDir
+_make depend
+_make
+_make install
+
+cd $depsSrcDir
+if [[ ! -d $curlSrcDir ]]; then
+    git clone https://github.com/curl/curl
+fi
+cd $curlSrcDir
+git checkout $curlTag
+./buildconf
+./configure --prefix=$curlDir --with-ssl=$sslDir --with-zlib=$zlibDir \
+            --host=x86_64-w64-mingw32
+_make
+_make install
+
+
+cd $depsSrcDir
+if [[ ! -d $boostSrcDir ]]; then
+    wget -qO- $boostUrl | tar xz
+fi
+
+cd $boostSrcDir
+echo "using gcc : mingw32 : x86_64-w64-mingw32-g++-posix ;" > user-config.jam
+
+# Workaround for https://github.com/boostorg/thread/issues/156
+# Older versions of mingw provided a different pthread lib.
+sed -i 's/lib$(libname)GC2.a/lib$(libname).a/g' ./libs/thread/build/Jamfile.v2
+sed -i 's/mthreads/pthreads/g' ./tools/build/src/tools/gcc.py
+sed -i 's/mthreads/pthreads/g' ./tools/build/src/tools/gcc.jam
+
+sed -i 's/pthreads/mthreads/g' ./tools/build/src/tools/gcc.py
+sed -i 's/pthreads/mthreads/g' ./tools/build/src/tools/gcc.jam
+
+export PTW32_INCLUDE=/usr/share/mingw-w64/include
+export PTW32_LIB=/usr/x86_64-w64-mingw32/lib
+
+# Fix getting Windows page size
+# TODO: send this upstream and maybe use a fork until it merges
+cat | patch -N boost/thread/pthread/thread_data.hpp <<EOL
+--- boost/thread/pthread/thread_data.hpp        2019-10-11 15:26:15.678703586 +0300
++++ boost/thread/pthread/thread_data.hpp.new    2019-10-11 15:26:07.321463698 +0300
+@@ -32,6 +32,10 @@
+ # endif
+ #endif
+
++#if defined(_WIN32)
++#include <windows.h>
++#endif
++
+ #include <pthread.h>
+ #include <unistd.h>
+
+@@ -54,6 +58,10 @@
+           if (size==0) return;
+ #ifdef BOOST_THREAD_USES_GETPAGESIZE
+           std::size_t page_size = getpagesize();
++#elif _WIN32
++          SYSTEM_INFO system_info;
++          ::GetSystemInfo (&system_info);
++          std::size_t page_size = system_info.dwPageSize;
+ #else
+           std::size_t page_size = ::sysconf( _SC_PAGESIZE);
+ #endif
+EOL
+
+# Use pthread if requested
+cat | patch -N boost/asio/detail/thread.hpp <<EOL
```

Comment:
is this patch still necessary if we are using boost 1.72? if yes, could you move it into `ceph/cmake/modules` and patch it if on win32 ? see also e461bda6c9507b6b30e574f35fdf904870363111 .

Reply:
Those patches are still needed unfortunately. We're not using that cmake module at the moment, one reason being that it will need some further changes in order to work with mingw. We can do that, but that means further straying away from the upstream version.

For now, I'd go with a simple inline TODO, something like "TODO: until upstream Boost gets patched, consider moving this to ceph/cmake/modules/BuildBoost.cmake and/or the upstream cmake module, which will first have to be updated in order to support mingw though.".

Reply:
sure. fine by me. would be great if these patches get upstreamed sooner!